In [34]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [35]:
data=pd.read_csv(r"/kaggle/input/real-or-fake-fake-jobposting-prediction/fake_job_postings.csv")

In [36]:
data.head()

job_id                                      title            location  \
0       1                           Marketing Intern    US, NY, New York   
1       2  Customer Service - Cloud Video Production      NZ, , Auckland   
2       3    Commissioning Machinery Assistant (CMA)       US, IA, Wever   
3       4          Account Executive - Washington DC  US, DC, Washington   
4       5                        Bill Review Manager  US, FL, Fort Worth   

  department salary_range                                    company_profile  \
0  Marketing          NaN  We're Food52, and we've created a groundbreaki...   
1    Success          NaN  90 Seconds, the worlds Cloud Video Production ...   
2        NaN          NaN  Valor Services provides Workforce Solutions th...   
3      Sales          NaN  Our passion for improving quality of life thro...   
4        NaN          NaN  SpotSource Solutions LLC is a Global Human Cap...   

                                         description  \
0  Food52, a fast-growing, James Beard Award-winn...   
1  Organised - Focused - Vibrant - Awesome!Do you...   
2  Our client, located in Houston, is actively se...   
3  THE COMPANY: ESRI – Environmental Systems Rese...   
4  JOB TITLE: Itemization Review ManagerLOCATION:...   

                                        requirements  \
0  Experience with content management systems a m...   
1  What we expect from you:Your key responsibilit...   
2  Implement pre-commissioning and commissioning ...   
3  EDUCATION: Bachelor’s or Master’s in GIS, busi...   
4  QUALIFICATIONS:RN license in the State of Texa...   

                                            benefits  telecommuting  \
0                                                NaN              0   
1  What you will get from usThrough being part of...              0   
2                                                NaN              0   
3  Our culture is anything but corporate—we have ...              0   
4                              Full Benefits Offered              0   

   has_company_logo  has_questions employment_type required_experience  \
0                 1              0           Other          Internship   
1                 1              0       Full-time      Not Applicable   
2                 1              0             NaN                 NaN   
3                 1              0       Full-time    Mid-Senior level   
4                 1              1       Full-time    Mid-Senior level   

  required_education                   industry              function  \
0                NaN                        NaN             Marketing   
1                NaN  Marketing and Advertising      Customer Service   
2                NaN                        NaN                   NaN   
3  Bachelor's Degree          Computer Software                 Sales   
4  Bachelor's Degree     Hospital & Health Care  Health Care Provider   

   fraudulent  
0           0  
1           0  
2           0  
3           0  
4           0

In [37]:
data.sample(5)

job_id                                         title  \
14071   14072               Customer Support Representative   
11885   11886            Software Developer .Net SQL Server   
5479     5480               Customer Service Representative   
7790     7791                               Program Manager   
5791     5792  Headhunter / Recruiter - Looking for Partner   

                   location            department salary_range  \
14071         GR, I, Athens  Business Development          NaN   
11885       US, VA, Fairfax                   NaN          NaN   
5479   ZA, GT, Johannesburg      Customer Service   5000-10000   
7790      US, CA, San Mateo                   NaN          NaN   
5791        US, VA, Ashburn                   NaN          NaN   

                                         company_profile  \
14071  Taxibeat is changing the global taxi industry....   
11885  FiLo is a leader in custom software applicatio...   
5479                                                 NaN   
7790   Tile is one of the most successful crowd-funde...   
5791   HireHopes is the premier professional services...   

                                             description  \
14071  Taxibeat is a new mobile service that aims to ...   
11885  Strong Experience writing and designing softwa...   
5479   Have you helped old ladies with their shopping...   
7790   The RoleYou are the hub of all the cross-funct...   
5791   HireHopes Consulting is looking for proven rec...   

                                            requirements  \
14071  As a Taxibeat support professional, you must:b...   
11885  Ability to acquire Secret level security clear...   
5479   What are the basic requirements?- Matric or eq...   
7790   B.S. in a technical discipline or equivalent e...   
5791   We are not looking for employees, we are looki...   

                                                benefits  telecommuting  \
14071  COMPENSATION:We offer basic salary, which prog...              0   
11885  Health Benefits Package OfferedLife Insurance ...              0   
5479                                                 NaN              0   
7790   Own equity in the company. Every employee is a...              0   
5791                                                 NaN              0   

       has_company_logo  has_questions employment_type required_experience  \
14071                 1              1             NaN                 NaN   
11885                 1              1       Full-time    Mid-Senior level   
5479                  0              1       Full-time         Entry level   
7790                  1              0       Full-time           Associate   
5791                  1              1       Full-time    Mid-Senior level   

              required_education              industry              function  \
14071                        NaN                   NaN                   NaN   
11885                        NaN     Computer Software                   NaN   
5479   High School or equivalent                   NaN      Customer Service   
7790           Bachelor's Degree  Consumer Electronics            Management   
5791                 Unspecified       Human Resources  Business Development   

       fraudulent  
14071           0  
11885           0  
5479            0  
7790            0  
5791            0

In [38]:
data.shape

(17880, 18)

In [39]:
from ydata_profiling import ProfileReport


In [40]:
prof=ProfileReport(data)
prof.to_file(output_file='output.html')

Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]


100%|██████████| 18/18 [00:08<00:00,  2.19it/s][A


Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

In [41]:
drop_col = [
    "job_id",
    "location",
    "title",
    "company_profile",
    "description",
    "requirements",
    "benefits",
    "telecommuting",
    "function",
    "industry"
]

df = data.drop(columns=drop_col)

In [42]:
df.head()

department salary_range  has_company_logo  has_questions employment_type  \
0  Marketing          NaN                 1              0           Other   
1    Success          NaN                 1              0       Full-time   
2        NaN          NaN                 1              0             NaN   
3      Sales          NaN                 1              0       Full-time   
4        NaN          NaN                 1              1       Full-time   

  required_experience required_education  fraudulent  
0          Internship                NaN           0  
1      Not Applicable                NaN           0  
2                 NaN                NaN           0  
3    Mid-Senior level  Bachelor's Degree           0  
4    Mid-Senior level  Bachelor's Degree           0

In [43]:
allowed_departments = {
    "sales",
    "engineering",
    "marketing",
    "operations",
    "development",
    "it",
    "product"
}


df["department"] = (
    df["department"]
    .fillna("unknown")
    .astype(str)
    .str.lower()
    .str.strip()
)

df["department"] = df["department"].apply(
    lambda x: x if x in allowed_departments else "other"
)

print(df["department"].value_counts(normalize=True))

department
other          0.871924
sales          0.032550
engineering    0.028635
marketing      0.023546
operations     0.015436
it             0.012919
development    0.008557
product        0.006432
Name: proportion, dtype: float64


In [44]:
df.head()

department salary_range  has_company_logo  has_questions employment_type  \
0  marketing          NaN                 1              0           Other   
1      other          NaN                 1              0       Full-time   
2      other          NaN                 1              0             NaN   
3      sales          NaN                 1              0       Full-time   
4      other          NaN                 1              1       Full-time   

  required_experience required_education  fraudulent  
0          Internship                NaN           0  
1      Not Applicable                NaN           0  
2                 NaN                NaN           0  
3    Mid-Senior level  Bachelor's Degree           0  
4    Mid-Senior level  Bachelor's Degree           0

In [45]:
df["salary_range"] = (
    df["salary_range"]
    .fillna("unknown")
    .astype(str)
    .str.strip()
    .str.lower()
)

invalid_salary_values = {"unknown", "0-0", "0", ""}

df["has_salary_range"] = (
    ~df["salary_range"].isin(invalid_salary_values)
).astype(int)
df = df.drop(columns=["salary_range"])

In [46]:
df["employment_type"] = (
    df["employment_type"]
    .fillna("unknown")
    .astype(str)
)

In [47]:
df.sample(5)

department  has_company_logo  has_questions employment_type  \
13610        other                 1              0       Full-time   
5982         other                 0              1       Full-time   
14912  development                 1              0         unknown   
10275        other                 1              0         unknown   
1636         other                 0              0         unknown   

      required_experience required_education  fraudulent  has_salary_range  
13610    Mid-Senior level                NaN           0                 0  
5982       Not Applicable                NaN           0                 0  
14912                 NaN                NaN           0                 0  
10275                 NaN                NaN           0                 0  
1636                  NaN                NaN           0                 0

In [48]:
df["required_experience"] = (
    df["required_experience"]
    .fillna("missing")
    .astype(str)
    .str.lower()
    .str.strip()
)

In [49]:
df["experience_missing"] = (
    df["required_experience"] == "missing"
).astype(int)

df["experience_not_applicable"] = (
    df["required_experience"] == "not applicable"
).astype(int)


In [50]:
def map_experience(x):
    if x in ['internship']:
        return 'internship'
    if x in ["entry level", "associate"]:
        return "junior"
    if x in ["mid-senior level", "director", "executive"]:
        return "senior"
    return "unspecified"

df["required_experience"] = df["required_experience"].apply(map_experience)


In [51]:
df["required_education"] = (
    df["required_education"]
    .fillna("missing")
    .astype(str)
    .str.lower()
    .str.strip()
)


df["education_missing"] = (
    df["required_education"] == "missing"
).astype(int)

df["education_unspecified"] = (
    df["required_education"] == "unspecified"
).astype(int)

def map_education(value):
    if "bachelor" in value:
        return "bachelors"
    if "master" in value:
        return "masters"
    if "high school" in value:
        return "high School or equivalent"
    if value in ["missing", "unspecified"]:
        return "unspecified"
    return "other"

df["required_education"] = df["required_education"].apply(map_education)

In [53]:
prof=ProfileReport(df)
prof.to_file(output_file='cleaned_output.html')

Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]


100%|██████████| 12/12 [00:00<00:00, 65.09it/s][A


Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

In [54]:
df.duplicated().value_counts()

True     16575
False     1305
Name: count, dtype: int64

In [52]:
df.sample(5)

department  has_company_logo  has_questions employment_type  \
6702         other                 1              1       Full-time   
7520         other                 1              0       Full-time   
15354  development                 1              1       Full-time   
1828         other                 0              1         unknown   
7874         other                 1              1       Full-time   

      required_experience required_education  fraudulent  has_salary_range  \
6702               senior          bachelors           0                 0   
7520               junior        unspecified           0                 0   
15354              junior          bachelors           0                 1   
1828          unspecified        unspecified           0                 0   
7874               junior        unspecified           0                 0   

       experience_missing  experience_not_applicable  education_missing  \
6702                    0                          0                  0   
7520                    0                          0                  1   
15354                   0                          0                  0   
1828                    1                          0                  1   
7874                    0                          0                  1   

       education_unspecified  
6702                       0  
7520                       0  
15354                      0  
1828                       0  
7874                       0

In [57]:
x = df.drop(columns=["fraudulent"])
y = df["fraudulent"]

x = pd.get_dummies(x, drop_first=True)

In [58]:
x.head()

has_company_logo  has_questions  has_salary_range  experience_missing  \
0                 1              0                 0                   0   
1                 1              0                 0                   0   
2                 1              0                 0                   1   
3                 1              0                 0                   0   
4                 1              1                 0                   0   

   experience_not_applicable  education_missing  education_unspecified  \
0                          0                  1                      0   
1                          1                  1                      0   
2                          0                  1                      0   
3                          0                  0                      0   
4                          0                  0                      0   

   department_engineering  department_it  department_marketing  ...  \
0                   False          False                  True  ...   
1                   False          False                 False  ...   
2                   False          False                 False  ...   
3                   False          False                 False  ...   
4                   False          False                 False  ...   

   employment_type_Part-time  employment_type_Temporary  \
0                      False                      False   
1                      False                      False   
2                      False                      False   
3                      False                      False   
4                      False                      False   

   employment_type_unknown  required_experience_junior  \
0                    False                       False   
1                    False                       False   
2                     True                       False   
3                    False                       False   
4                    False                       False   

   required_experience_senior  required_experience_unspecified  \
0                       False                            False   
1                       False                             True   
2                       False                             True   
3                        True                            False   
4                        True                            False   

   required_education_high School or equivalent  required_education_masters  \
0                                         False                       False   
1                                         False                       False   
2                                         False                       False   
3                                         False                       False   
4                                         False                       False   

   required_education_other  required_education_unspecified  
0                     False                            True  
1                     False                            True  
2                     False                            True  
3                     False                           False  
4                     False                           False  

[5 rows x 26 columns]

In [59]:
from sklearn.model_selection import train_test_split

In [60]:
x_train,x_test,y_train,y_test=train_test_split(x,y,random_state=42,test_size=0.3,stratify=y)

In [65]:
from sklearn.linear_model import LogisticRegression

model = LogisticRegression(
    max_iter=1000,
    class_weight="balanced"
)

model.fit(x_train, y_train)

LogisticRegression(class_weight='balanced', max_iter=1000)

In [67]:
y_pred=model.predict(x_test)

In [68]:

from sklearn.metrics import accuracy_score
print(accuracy_score(y_test, y_pred))

0.7798284862043251


In [69]:
from sklearn.metrics import confusion_matrix
print(confusion_matrix(y_test, y_pred))

[[3989 1115]
 [  66  194]]


In [70]:
from xgboost import XGBClassifier

xgb = XGBClassifier(
    n_estimators=200,
    max_depth=5,
    learning_rate=0.1,
    subsample=0.8,
    colsample_bytree=0.8,
    scale_pos_weight=(y_train == 0).sum() / (y_train == 1).sum(),
    eval_metric="logloss",
    random_state=42
)

xgb.fit(x_train, y_train)

XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=0.8, device=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric='logloss',
              feature_types=None, feature_weights=None, gamma=None,
              grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=0.1, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=5, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              multi_strategy=None, n_estimators=200, n_jobs=None,
              num_parallel_tree=None, ...)

In [72]:
y_xgb=xgb.predict(x_test)

In [74]:
print(accuracy_score(y_test, y_xgb))
print(confusion_matrix(y_test, y_xgb))

0.8234526472781506
[[4211  893]
 [  54  206]]


| Metric | Logistic | XGBoost | Change  |
| ------ | -------- | ------- | ------- |
| TN     | 3989     | 4211    | ⬆️ +222 |
| FP     | 1115     | 893     | ⬇️ −222 |
| FN     | 66       | 54      | ⬇️ −12  |
| TP     | 194      | 206     | ⬆️ +12  |


In [76]:
import joblib

joblib.dump(model, "logistic_baseline.pkl")
joblib.dump(xgb, "xgboost_model.pkl")

['xgboost_model.pkl']

In [77]:
joblib.dump(x_train.columns.tolist(), "feature_columns.pkl")

['feature_columns.pkl']

In [79]:
import numpy as np

p = xgb.predict_proba(x_test)[:, 1]

df_meta = pd.DataFrame({
    "predicted_label": (p >= 0.5).astype(int),
    "predicted_probability": p,
    "confidence_margin": np.abs(p - 0.5),
    "entropy": -(p * np.log(p + 1e-8) + (1 - p) * np.log(1 - p + 1e-8)),
    "true_label": y_test.values
})

In [80]:
df_meta.head()

predicted_label  predicted_probability  confidence_margin   entropy  \
0                1               0.698246           0.198246  0.612343   
1                0               0.018877           0.481123  0.093637   
2                0               0.000372           0.499628  0.003311   
3                0               0.189014           0.310986  0.484790   
4                0               0.028761           0.471239  0.130408   

   true_label  
0           0  
1           0  
2           0  
3           0  
4           0

In [81]:
df_meta["is_correct"] = (
    df_meta["predicted_label"] == df_meta["true_label"]
).astype(int)

entropy_threshold = df_meta["entropy"].quantile(0.75)

df_meta["low_confidence"] = (
    df_meta["entropy"] >= entropy_threshold
).astype(int)


df_meta["needs_human_review"] = (
    (df_meta["is_correct"] == 0) |
    (df_meta["low_confidence"] == 1)
).astype(int)

In [82]:
df_meta.head()

predicted_label  predicted_probability  confidence_margin   entropy  \
0                1               0.698246           0.198246  0.612343   
1                0               0.018877           0.481123  0.093637   
2                0               0.000372           0.499628  0.003311   
3                0               0.189014           0.310986  0.484790   
4                0               0.028761           0.471239  0.130408   

   true_label  is_correct  low_confidence  needs_human_review  
0           0           0               1                   1  
1           0           1               0                   0  
2           0           1               0                   0  
3           0           1               0                   0  
4           0           1               0                   0

In [83]:
df_meta_B = df_meta.drop(columns=["true_label", "is_correct", "low_confidence"])

In [84]:
df_meta_B.sample(5)

predicted_label  predicted_probability  confidence_margin   entropy  \
2705                0               0.020631           0.479369  0.100483   
2697                0               0.004295           0.495705  0.027694   
1213                1               0.698246           0.198246  0.612343   
195                 0               0.010041           0.489959  0.056190   
4683                1               0.512292           0.012292  0.692845   

      needs_human_review  
2705                   0  
2697                   0  
1213                   1  
195                    0  
4683                   1

In [85]:
df_meta_B.shape

(5364, 5)

In [86]:
df_meta_B["needs_human_review"].value_counts(normalize=True)

needs_human_review
0    0.634601
1    0.365399
Name: proportion, dtype: float64

In [87]:

X_B = df_meta_B.drop(columns=["needs_human_review"])
y_B = df_meta_B["needs_human_review"]

X_B_train, X_B_val, y_B_train, y_B_val = train_test_split(
    X_B,
    y_B,
    test_size=0.3,
    stratify=y_B,
    random_state=42
)

In [88]:
from xgboost import XGBClassifier

model_B = XGBClassifier(
    n_estimators=120,
    max_depth=3,
    learning_rate=0.1,
    subsample=0.8,
    colsample_bytree=0.8,
    eval_metric="logloss",
    random_state=42
)

model_B.fit(X_B_train, y_B_train)

XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=0.8, device=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric='logloss',
              feature_types=None, feature_weights=None, gamma=None,
              grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=0.1, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=3, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              multi_strategy=None, n_estimators=120, n_jobs=None,
              num_parallel_tree=None, ...)

In [89]:
from sklearn.metrics import classification_report, confusion_matrix

y_B_prob = model_B.predict_proba(X_B_val)[:, 1]
y_B_pred = (y_B_prob >= 0.5).astype(int)

print(confusion_matrix(y_B_val, y_B_pred))
print(classification_report(y_B_val, y_B_pred))

[[983  39]
 [ 14 574]]
              precision    recall  f1-score   support

           0       0.99      0.96      0.97      1022
           1       0.94      0.98      0.96       588

    accuracy                           0.97      1610
   macro avg       0.96      0.97      0.96      1610
weighted avg       0.97      0.97      0.97      1610



In [90]:
accuracy_score(y_B_val,y_B_pred)

0.9670807453416149

In [91]:
import numpy as np

for t in [0.3, 0.4, 0.5, 0.6]:
    preds = (y_B_prob >= t).astype(int)
    fn = ((preds == 0) & (y_B_val == 1)).sum()
    print(f"Threshold {t}: False negatives = {fn}")

Threshold 0.3: False negatives = 13
Threshold 0.4: False negatives = 13
Threshold 0.5: False negatives = 14
Threshold 0.6: False negatives = 14


In [93]:
joblib.dump(model_B, "model_b_risk_classifier.pkl")

['model_b_risk_classifier.pkl']

In [94]:
joblib.dump(
    X_B.columns.tolist(),
    "model_b_features.pkl"
)

['model_b_features.pkl']

In [97]:
eval_summary = {
    "false_negatives": int(((y_B_pred == 0) & (y_B_val == 1)).sum()),
    "false_positives": int(((y_B_pred == 1) & (y_B_val == 0)).sum()),
    "threshold": 0.5,
    "recall_risky": float(
        (y_B_pred[y_B_val == 1] == 1).mean()
    )
}

joblib.dump(
    eval_summary,
    "model_b_eval_summary.pkl"
)

['model_b_eval_summary.pkl']